mp/src/materialsystem/stdshaders/pbr_dx9.cpp

```diff
@@ -0,0 +1,524 @@
+            }
+
+            // if the hardware doesn't support cubemaps, use spheremaps instead
+            if (!g_pHardwareConfig->SupportsCubeMaps())
+            {
+                SET_FLAGS(MATERIAL_VAR_ENVMAPSPHERE);
+            }
+
+            // done like this so the user could set $envmapsphere themselves if they wanted to
+            if (!IS_FLAG_SET(MATERIAL_VAR_ENVMAPSPHERE))
+            {
+                LoadCubeMap(info.Envmap, flags);
+            }
+            else
+            {
+                LoadTexture(info.Envmap, flags);
+            }
+        }
+
+        // setting up flashlight texture
+        if (info.FlashlightTexture != -1 && params[info.FlashlightTexture]->IsDefined())
+        {
+            LoadTexture(info.FlashlightTexture, TEXTUREFLAGS_SRGB); // flashlight texture is srgb
+        }
+    };
+
+    // drawing the shader
+    SHADER_DRAW
+    {
+        PBR_DX9_Vars_t info;
+        SetupVars(info);
+
+        // setting up some booleans
+        bool bIsAlphaTested     = IS_FLAG_SET(MATERIAL_VAR_ALPHATEST) != 0;
+        bool bHasAlbedo         = (info.Albedo != -1) && params[info.Albedo]->IsTexture();
+        bool bHasNormal         = (info.Normal != -1) && params[info.Normal]->IsTexture();
+        bool bHasMrao           = (info.Mrao != -1) && params[info.Mrao]->IsTexture();
+        bool bHasEmissive       = (info.Emissive != -1) && params[info.Emissive]->IsTexture();
+        bool bHasEnvmap         = (info.Envmap != -1) && params[info.Envmap]->IsTexture();
+        bool bHasFlashlight     = UsingFlashlight(params);
+        bool bLightmapped       = IS_FLAG_SET(MATERIAL_VAR_MODEL) == 0;
+
+        // determining whether we're dealing with a fully opaque material
+        BlendType_t nBlendType = this->EvaluateBlendRequirements(info.Albedo, true);
+        bool bFullyOpaque = (nBlendType != BT_BLENDADD) && (nBlendType != BT_BLEND) && !bIsAlphaTested;
+
+        if (this->IsSnapshotting()) // shadow state
+        {
+            // if alphatest is on, enable it
+            pShaderShadow->EnableAlphaTest(bIsAlphaTested);
+
+            // if the user set an alpha reference, use it
+            if (info.AlphaTestReference != -1 && params[info.AlphaTestReference]->GetFloatValue() > 0.0f)
+            {
+                pShaderShadow->AlphaFunc(SHADER_ALPHAFUNC_GEQUAL, params[info.AlphaTestReference]->GetFloatValue());
+            }
+
+            // if the flashlight is on, get the shadow filter mode
+            int nShadowFilterMode = bHasFlashlight ? g_pHardwareConfig->GetShadowFilterMode() : 0;
+
+            // if the albedo texture exists, set the default blendding shadow state
+            if (params[info.Albedo]->IsTexture())
+            {
+                SetDefaultBlendingShadowState(info.Albedo, true);
+            }
+
+            // setting up samplers
+            pShaderShadow->EnableTexture(SHADER_SAMPLER0, true);    // Albedo Texture
```

Comment:
Perhaps make some defines up top that map names to sampler slots? E.g. `#define ALBEDO_SAMPLER SHADER_SAMPLER0`.

Reply:
good idea